### PR TITLE
Agents SPA

### PR DESCRIPTION
**Pull request by GaneshSPatil**
The new agents page as a single page application using mithril

### CODE REVIEW

File: server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/views/agents/environments_list_widget.js.msx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+          var checkedAgents = _.compact(args.selectAllViewModel.checkedAgents());
+          return _.reduce(checkedAgents, function (environments, agent) {
+            environments[agent.hostname()] = agent.environments();
+            return environments;
+          }, {});
+        };
+
+        this.updateEnvironments = function () {
+          args.togglePolling(false);
+          $('#environments_list').removeClass('is-open');
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/views/agents/environments_list_widget.js.msx]: Please don't do this. You're mutating virtual dom element from another using jQuery. Ideally you'd want the `#environments_list` element's `class` to be set via a function that's controlled by a flag.

**Comment by zabil** [comment]
I only see jasmine spec/tests for models.
Please make sure the view, controller of all the widgets are have tests

File: server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/helpers/mithril_component_mixins.js.msx [hunk 2]
```diff
@@ -27,6 +27,10 @@ define(['string-plus', 'mithril', 'lodash', 'helpers/form_helper'], function (s,
 
     vm = vm || {};
 
+    this.removeVMState = function (key) {
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/helpers/mithril_component_mixins.js.msx]: Looks like dead code

File: server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/models/agents/environments.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+define([], function () {
+  var Environments = {};
+
+    Environments.initializeWith = function (values) {
+        Environments.list = values || [];
+  };
+
+    Environments.list = [];
+
+  return Environments;
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/models/agents/environments.js]: Bad indent?

File: server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/single_page_apps/agents.js.msx [hunk 4]
```diff
@@ -0,0 +1,41 @@
+/*
```
Comment (maheshp) [inline server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/single_page_apps/agents.js.msx]: Not a msx file

**Comment by zabil** [comment]
For this and the other SPA's must set strict content security policies before it's too late. 
More details here.
https://content-security-policy.com/

I suggest we add these headers for the new rails controllers and then slowly roll it out to other pages. 
cc @maheshp

File: server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/views/agents/button_row_widget.js.msx [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        };
+
+
+        this.resources = [];
+        this.fetchResources = function() {
+          m.request({
+            method: 'GET',
+            url: Routes.apiv1AdminInternalResourcesPath(),
+            config: mrequest.xhrConfig.v1,
+            //async: false,
```
Comment (zabil) [inline server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/views/agents/button_row_widget.js.msx]: Remove all these and debugger comments

File: server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/models/agents/environments.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    return _.every(selectedAgents, function (agent) {
+      return _.includes(agent.environments(), value);
+    })
+  };
+
+  var isPresentOnAny = function (selectedAgents, value) {
+    return _.some(selectedAgents, function (agent) {
+      return _.includes(agent.environments(), value);
+    })
+  };
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/models/agents/environments.js]: If you extract `.environments()` and `.resources()` (from the file below) you may be able to extract `isPresentOn..` methods into a common codebase (probably inside `tri_state_checkbox.js`)

File: server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/single_page_apps/agents.js.msx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  'views/agents/agents_widget',
+  'views/agents/agent_spa',
+  'foundation.util.mediaQuery',
+  'foundation.dropdownMenu',
+  'foundation.responsiveToggle',
+  'foundation.dropdown'
+], function ($, m, _,
+             f, ComponentMixins,
+             JsRoutes,
+             Agents, Resources, Environments, AgentsWidget, AgentsSPA) {
```
Comment (ketan) [inline server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/single_page_apps/agents.js.msx]: Most of these components are unused `_`, `f`, `ComponentMixins`, `Agents`, `AgentsWidget`, isn't the IDE telling you about these?

**Pull request merged**